keanu-python/keanu/sequence.py

```diff
@@ -59,17 +62,30 @@ def __init__(self,
                 "Cannot create a sequence of an unknown size: you must specify either a count of a data_generator")
         elif count is not None and data_generator is not None:
             raise ValueError("If you pass in a data_generator you cannot also pass in a count")
+        elif factory is None and factories is None:
+            raise ValueError("You must provide a value for either the 'factory' or 'factories' input")
 
         if count is not None:
-            function = lambda p: factory(SequenceItem(p))
-            consumer = Consumer(function)
-            builder = builder.count(count).withFactory(consumer)
+            if factories is not None:
+                functions = [Consumer(lambda p, f=f: f(SequenceItem(p))) for f in factories]
+                java_functions = ListConverter().convert(functions, k._gateway._gateway_client)
+                builder = builder.count(count).withFactories(java_functions)
+            elif factory is not None:
+                function = lambda p: factory(SequenceItem(p))
+                consumer = Consumer(function)
+                builder = builder.count(count).withFactory(consumer)
 
         if data_generator is not None:
-            bifunction = lambda p, data: factory(SequenceItem(p), data)
-            biconsumer = BiConsumer(bifunction)
-            data_generator_java = (k.to_java_map(m) for m in data_generator)
-            builder = builder.fromIterator(JavaIterator(data_generator_java)).withFactory(biconsumer)
+            if factories is not None:
+                bifunctions = [Consumer(lambda p, data: factory(SequenceItem(p), data)) for factory in factories]
```

Comment:
presumably this also suffers from the "Late Binding Closures" gotcha?

Reply:
Yep - have used the `partial` approach on this one now too :)